modules/javafx.graphics/src/main/java/javafx/scene/layout/HBox.java

```diff
@@ -405,139 +410,164 @@ private boolean shouldFillHeight() {
-            adjustAreaWidths(managed, prefWidths, width, -1);
-            contentHeight = computeMaxMinAreaHeight(managed, marginAccessor, prefWidths[0], getAlignmentInternal().getVpos());
+            double[] actualAreaWidths = distributeSpace(managed, width, -1);
+
+            contentHeight = computeMaxMinAreaHeight(managed, marginAccessor, actualAreaWidths, getAlignmentInternal().getVpos());
         } else {
             contentHeight = computeMaxMinAreaHeight(managed, marginAccessor, getAlignmentInternal().getVpos());
         }
-        return snapSpaceY(insets.getTop()) +
-               contentHeight +
-               snapSpaceY(insets.getBottom());
+
+        return snapSpaceY(
+            snapSpaceY(insets.getTop()) +
+            contentHeight +
+            snapSpaceY(insets.getBottom())
+        );
     }
 
     @Override protected double computePrefWidth(double height) {
         Insets insets = getInsets();
-        return snapSpaceX(insets.getLeft()) +
-               computeContentWidth(getManagedChildren(), height, false) +
-               snapSpaceX(insets.getRight());
+
+        return snapSpaceX(
+            snapSpaceX(insets.getLeft()) +
+            computeContentWidth(getManagedChildren(), height, false) +
+            snapSpaceX(insets.getRight())
+        );
     }
 
     @Override protected double computePrefHeight(double width) {
         Insets insets = getInsets();
         List<Node>managed = getManagedChildren();
         double contentHeight = 0;
         if (width != -1 && getContentBias() != null) {
-            double prefWidths[][] = getAreaWidths(managed, -1, false);
-            adjustAreaWidths(managed, prefWidths, width, -1);
-            contentHeight = computeMaxPrefAreaHeight(managed, marginAccessor, prefWidths[0], getAlignmentInternal().getVpos());
+            double[] actualAreaWidths = distributeSpace(managed, width, -1);
+
+            contentHeight = computeMaxPrefAreaHeight(managed, marginAccessor, actualAreaWidths, getAlignmentInternal().getVpos());
         } else {
             contentHeight = computeMaxPrefAreaHeight(managed, marginAccessor, getAlignmentInternal().getVpos());
         }
-        return snapSpaceY(insets.getTop()) +
-               contentHeight +
-               snapSpaceY(insets.getBottom());
+
+        return snapSpaceY(
+            snapSpaceY(insets.getTop()) +
+            contentHeight +
+            snapSpaceY(insets.getBottom())
+        );
     }
 
-    private double[][] getAreaWidths(List<Node>managed, double height, boolean minimum) {
-        // height could be -1
-        double[][] temp = getTempArray(managed.size());
-        final double insideHeight = height == -1? -1 : height -
-                                     snapSpaceY(getInsets().getTop()) - snapSpaceY(getInsets().getBottom());
-        final boolean shouldFillHeight = shouldFillHeight();
-        for (int i = 0, size = managed.size(); i < size; i++) {
+    private double getSnapScaleX() {
```

Comment:
this method in Region should have been made protected to avoid re-implementation...